[DDP] Test inference works with eval() and no_grad()

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59666 [DDP] Test inference works with eval() and no_grad()**

Tests that inference with DDP model won't hang when user sets eval()
or no_grad(). Note that if the model has a syncBN layer, they need both eval()
and no_grad() as eval() makes SyncBN work like a regular BN layer.

Differential Revision: [D28974146](https://our.internmc.facebook.com/intern/diff/D28974146/)

rohan-varma committed:
[DDP] Test inference works with eval() and no_grad()

Tests that inference with DDP model won't hang when user sets eval()
or no_grad(). Note that if the model has a syncBN layer, they need both eval()
and no_grad() as eval() makes SyncBN work like a regular BN layer.

Differential Revision: [D28974146](https://our.internmc.facebook.com/intern/diff/D28974146/)

[ghstack-poisoned]

rohan-varma commented:
cc @mrshenli

rohan-varma committed:
Update on "[DDP] Test inference works with eval() and no_grad()"


Tests that inference with DDP model won't hang when user sets eval()
or no_grad(). Note that if the model has a syncBN layer, they need both eval()
and no_grad() as eval() makes SyncBN work like a regular BN layer.

Differential Revision: [D28974146](https://our.internmc.facebook.com/intern/diff/D28974146/)

[ghstack-poisoned]

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6943,10 +6943,65 @@ def test_ddp_multiple_nested_unused_params_err_ignore_params(self):
+            if self.rank == 0:
+                with torch.no_grad():
+                    for _ in range(6):
+                        ddp_out = model(inp)
```
Nit: If you rename `model` as `ddp_model`, then you can just use one line here:
`self.assertEqual(ddp_model(inp), local_model(inp))`

Just more concise.

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6943,10 +6943,65 @@ def test_ddp_multiple_nested_unused_params_err_ignore_params(self):
+            # or eval setting and there is no hang.
+            rank = self.rank
+            torch.cuda.set_device(rank)
+            model = Net().cuda()
```
Nit: I believe you can just create a for loop over two tuples of <model, input>, where the models are `Net().cuda()` and `nn.SyncBatchNorm`. This can save some duplicate code and improve the readability.

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6943,10 +6943,65 @@ def test_ddp_multiple_nested_unused_params_err_ignore_params(self):
+                    self.assertEqual(ddp_out, local_out)
+                torch.cuda.synchronize()
+
+            self._barrier(timeout=30)
```
Are all the 3 barriers here necessary, even after `cuda.sync`? Why do they need a non-default higher timeout here?

rohan-varma replied:
I don't think the calls to `synchronize` are necessary actually, will remove those.

The calls to barrier are there because we're only running inference on rank 0, if the inference unexpectedly takes too log, the default barrier which I believe has a timeout of 10s can timeout, leading to a false positive failure.

In this test the inference does not take nearly the full 30s but wanted to have plenty of buffer to avoid flakiness.

rohan-varma committed:
Update on "[DDP] Test inference works with eval() and no_grad()"


Tests that inference with DDP model won't hang when user sets eval()
or no_grad(). Note that if the model has a syncBN layer, they need both eval()
and no_grad() as eval() makes SyncBN work like a regular BN layer.

Differential Revision: [D28974146](https://our.internmc.facebook.com/intern/diff/D28974146/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Test inference works with eval() and no_grad()"


Tests that inference with DDP model won't hang when user sets eval()
or no_grad(). Note that if the model has a syncBN layer, they need both eval()
and no_grad() as eval() makes SyncBN work like a regular BN layer.

Differential Revision: [D28974146](https://our.internmc.facebook.com/intern/diff/D28974146/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Test inference works with eval() and no_grad()"


Tests that inference with DDP model won't hang when user sets eval()
or no_grad(). Note that if the model has a syncBN layer, they need both eval()
and no_grad() as eval() makes SyncBN work like a regular BN layer.

Differential Revision: [D28974146](https://our.internmc.facebook.com/intern/diff/D28974146/)

[ghstack-poisoned]

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6943,10 +6943,58 @@ def test_ddp_multiple_nested_unused_params_err_ignore_params(self):
+                device_ids=[rank]
+            )
+            inp = torch.randn(10, 2, device=rank)
+            inp_syncbn = torch.randn(10, 2, 4, 4, device=rank)
```
What's the abbreviation of "inp"?

rohan-varma replied:
it is short for "input"

rohan-varma committed:
Update on "[DDP] Test inference works with eval() and no_grad()"


Tests that inference with DDP model won't hang when user sets eval()
or no_grad(). Note that if the model has a syncBN layer, they need both eval()
and no_grad() as eval() makes SyncBN work like a regular BN layer.

Differential Revision: [D28974146](https://our.internmc.facebook.com/intern/diff/D28974146/)

[ghstack-poisoned]
